src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -340,6 +340,121 @@ impl ScalarFunc {
         };
         Ok(Some(Cow::Owned(r.into_bytes())))
     }
+
+    #[inline]
+    pub fn substring_2_args<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        let pos = try_opt!(self.children[1].eval_int(ctx, row));
+        if pos == 0 {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+
+        let start = if mysql::has_unsigned_flag(self.children[1].get_tp().get_flag()) {
+            find_start(pos as u64 as usize, s.char_indices())
+        } else if pos > 0 {
+            find_start(pos as usize, s.char_indices())
+        } else {
+            let pos = 0_i64
+                .checked_sub(pos)
+                .map(|i| i as usize)
+                .unwrap_or_else(|| i64::max_value() as usize + 1);
+            find_start(pos, s.char_indices().rev())
+        };
+
+        if let Some(start) = start {
+            Ok(Some(Cow::Owned(s[start..].as_bytes().to_vec())))
+        } else {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+    }
+
+    #[inline]
+    pub fn substring_3_args<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let s = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        let pos = try_opt!(self.children[1].eval_int(ctx, row));
+        let len = try_opt!(self.children[2].eval_int(ctx, row));
+        let mut positive_search = true;
+
+        let pos = if mysql::has_unsigned_flag(self.children[1].get_tp().get_flag()) {
```

Comment:
the process of calculating `pos` seems to be the same as in `substring_2_args`. Is it possible to extract a function? Or delegate that function to call this function?

Reply:
I extracted a function, but not found a better name.  please take a look